File: php/src/Core/Headers.php
<?php

declare(strict_types=1);

namespace SolanaMpp\Core;

use InvalidArgumentException;

/**
 * Formats and parses MPP HTTP authentication headers.
 */
final class Headers
{
    public const PAYMENT_SCHEME = 'Payment';
    public const WWW_AUTHENTICATE = 'www-authenticate';
    public const AUTHORIZATION = 'authorization';
    public const PAYMENT_RECEIPT = 'payment-receipt';

    /**
     * Format a Payment challenge as a WWW-Authenticate header.
     */
    public static function formatWwwAuthenticate(Challenge $challenge): string
    {
        $parts = [
            sprintf('id="%s"', self::escapeQuoted($challenge->id)),
            sprintf('realm="%s"', self::escapeQuoted($challenge->realm)),
            sprintf('method="%s"', self::escapeQuoted($challenge->method)),
            sprintf('intent="%s"', self::escapeQuoted($challenge->intent)),
            sprintf('request="%s"', self::escapeQuoted($challenge->request)),
        ];
        if ($challenge->expires !== '') {
            $parts[] = sprintf('expires="%s"', self::escapeQuoted($challenge->expires));
        }
        if ($challenge->digest !== '') {
            $parts[] = sprintf('digest="%s"', self::escapeQuoted($challenge->digest));
        }
        if ($challenge->opaque !== null) {
            $parts[] = sprintf('opaque="%s"', self::escapeQuoted($challenge->opaque));
        }

        return self::PAYMENT_SCHEME . ' ' . implode(', ', $parts);
    }

    /**
     * Parse all Payment challenges across one or more WWW-Authenticate header values (RFC 7235 sec 4.1).
     *
     * @param iterable<string>|string $headers
     * @return array<int, Challenge>
     */
    public static function parseWwwAuthenticateAll(iterable|string $headers): array
    {
        $list = is_string($headers) ? [$headers] : $headers;
        $challenges = [];
        foreach ($list as $header) {
            foreach (self::splitPaymentChallengeValues($header) as $chunk) {
                $challenges[] = self::parseWwwAuthenticate($chunk);
            }
        }

        return $challenges;
    }

    /**
     * Split a header value into individual `Payment` challenge chunks (quote-aware).
     *
     * @return array<int, string>
     */
    private static function splitPaymentChallengeValues(string $header): array
    {
        $length = strlen($header);
        $starts = [];
        $inQuote = false;
        $escaped = false;
        $i = 0;
        $scheme = self::PAYMENT_SCHEME;
        $sLen = strlen($scheme);

        while ($i < $length) {
            $ch = $header[$i];
            if ($inQuote) {
                if ($escaped) {
                    $escaped = false;
                } elseif ($ch === '\\') {
                    $escaped = true;
                } elseif ($ch === '"') {
                    $inQuote = false;
                }
                $i++;
                continue;
            }
            if ($ch === '"') {
                $inQuote = true;
                $i++;
                continue;
            }
            if (self::isPaymentSchemeStart($header, $i, $scheme, $sLen, $length)) {
                $starts[] = $i;
                $i += $sLen;
                continue;
            }
            $i++;
        }

        if ($starts === []) {
            return [];
        }

        $chunks = [];
        foreach ($starts as $index => $start) {
            $end = $starts[$index + 1] ?? $length;
            $chunk = trim(substr($header, $start, $end - $start));
            $chunk = rtrim($chunk, ', ');
            if ($chunk !== '') {
                $chunks[] = $chunk;
            }
        }
        return $chunks;
    }

    private static function isPaymentSchemeStart(string $header, int $index, string $scheme, int $sLen, int $length): bool
    {
        if ($index + $sLen >= $length) {
            return false;
        }
        if (strcasecmp(substr($header, $index, $sLen), $scheme) !== 0) {
            return false;
        }
        $next = $header[$index + $sLen];
        if ($next !== ' ' && $next !== "\t") {
            return false;
        }
        $prev = $index - 1;
        while ($prev >= 0 && ($header[$prev] === ' ' || $header[$prev] === "\t")) {
            $prev--;
        }
        return $prev < 0 || $header[$prev] === ',';
    }

    /**
     * Parse a WWW-Authenticate header into a Payment challenge.
     */
    public static function parseWwwAuthenticate(string $header): Challenge
    {
        $payment = self::extractPaymentChallenge($header);
        if ($payment === null) {
            throw new InvalidArgumentException('Expected Payment scheme');
        }

        $params = self::parseAuthParams(trim(substr($payment, strlen(self::PAYMENT_SCHEME))));
        foreach (['id', 'realm', 'method', 'intent', 'request'] as $required) {
            if (($params[$required] ?? '') === '') {
                throw new InvalidArgumentException(sprintf('Missing "%s" field', $required));
            }
        }

        Base64Url::decodeJson($params['request']); // validate the encoded charge request

        return new Challenge(
            id: $params['id'],
            realm: $params['realm'],
            method: $params['method'],
            intent: $params['intent'],
            request: $params['request'],
            expires: $params['expires'] ?? '',
            digest: $params['digest'] ?? '',
            opaque: $params['opaque'] ?? null,
        );
    }

    /**
     * Format a receipt as an unpadded base64url payment-receipt header.
     */
    public static function formatReceipt(Receipt $receipt): string
    {
        return Base64Url::encodeJson($receipt->toArray());
    }

    /**
     * Parse a payment-receipt header into a receipt object.
     */
    public static function parseReceipt(string $header): Receipt
    {
        if (strlen($header) > 16 * 1024) {
            throw new InvalidArgumentException('Receipt exceeds maximum length of 16384 bytes');
        }

        return Receipt::fromArray(Base64Url::decodeJson(trim($header)));
    }

    private static function extractPaymentChallenge(string $header): ?string
    {
        $lower = strtolower($header);
        $needle = strtolower(self::PAYMENT_SCHEME . ' ');
        $position = strpos($lower, $needle);
        if ($position === false) {
            return null;
        }

        return trim(substr($header, $position));
    }

    /**
     * @return array<string, string>
     */
    private static function parseAuthParams(string $value): array
    {
        $params = [];
        $length = strlen($value);
        $index = 0;

        while ($index < $length) {
            while ($index < $length && ($value[$index] === ' ' || $value[$index] === "\t" || $value[$index] === ',')) {
                $index++;
            }
            if ($index >= $length) {
                break;
            }

            $keyStart = $index;
            while ($index < $length && $value[$index] !== '=' && $value[$index] !== ',' && $value[$index] !== ' ' && $value[$index] !== "\t") {
                $index++;
            }
            $key = substr($value, $keyStart, $index - $keyStart);
            while ($index < $length && ($value[$index] === ' ' || $value[$index] === "\t")) {
                $index++;
            }
            if ($key === '' || $index >= $length || $value[$index] !== '=') {
                throw new InvalidArgumentException('Invalid auth parameter');
            }
            $index++;
            while ($index < $length && ($value[$index] === ' ' || $value[$index] === "\t")) {
                $index++;
            }

            if ($index < $length && $value[$index] === '"') {
                [$parsed, $index] = self::parseQuotedValue($value, $index + 1);
                if (array_key_exists($key, $params)) {
                    throw new InvalidArgumentException('Duplicate auth parameter');
                }
                $params[$key] = $parsed;
                continue;
            }

            $valueStart = $index;
            while ($index < $length && $value[$index] !== ',') {
                $index++;
            }
            if (array_key_exists($key, $params)) {
                throw new InvalidArgumentException('Duplicate auth parameter');
            }
            $params[$key] = trim(substr($value, $valueStart, $index - $valueStart));
        }

        return $params;
    }

    /**
     * @return array{0: string, 1: int}
     */
    private static function parseQuotedValue(string $value, int $index): array
    {
        $buffer = '';
        $length = strlen($value);
        while ($index < $length) {
            $char = $value[$index];
            if ($char === '\\') {
                $index++;
                if ($index >= $length) {
                    throw new InvalidArgumentException('Invalid quoted value');
                }
                $buffer .= $value[$index];
                $index++;
                continue;
            }
            if ($char === '"') {
                return [$buffer, $index + 1];
            }
            $buffer .= $char;
            $index++;
        }

        throw new InvalidArgumentException('Unterminated quoted value');
    }

    private static function escapeQuoted(string $value): string
    {
        if (str_contains($value, "\r") || str_contains($value, "\n")) {
            throw new InvalidArgumentException('Invalid header value');
        }

        return str_replace(['\\', '"'], ['\\\\', '\\"'], $value);
    }
}

File: python/src/solana_pay_kit/protocols/mpp/core/headers.py
"""Header parsing and formatting for Web Payment Auth.

No regex for auth-param parsing -- follows the same hand-rolled parser as the
Rust and Go implementations.
"""

from __future__ import annotations

import json
import re
from collections.abc import Iterable
from typing import Any

from solana_pay_kit.protocols.mpp.core.base64url import decode, decode_json, encode_json
from solana_pay_kit.protocols.mpp.core.types import ChallengeEcho, PaymentChallenge, PaymentCredential, Receipt

MAX_TOKEN_LEN = 16 * 1024

PAYMENT_SCHEME = "Payment"
WWW_AUTHENTICATE_HEADER = "www-authenticate"
AUTHORIZATION_HEADER = "authorization"
PAYMENT_RECEIPT_HEADER = "payment-receipt"


class ParseError(Exception):
    """Failed to parse a payment header."""


# ISO-8601 / RFC 3339 timestamp: YYYY-MM-DDTHH:MM:SS[.fff](Z|±HH:MM).
# Matches the canonical mpp-tools receipt timestamp validation; rejects loose
# forms like "Jan 29 2026 12:00".
_ISO8601_RE = re.compile(
    r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})$"
)


# ---------------------------------------------------------------------------
# WWW-Authenticate
# ---------------------------------------------------------------------------


def parse_www_authenticate(header: str) -> PaymentChallenge:
    """Parse a WWW-Authenticate header into a PaymentChallenge."""
    rest = _strip_payment_scheme(header)
    if rest is None:
        raise ParseError("Expected 'Payment' scheme")

    # Must have a space or tab separator
    if not rest or (rest[0] != " " and rest[0] != "\t"):
        raise ParseError("Expected space after 'Payment' scheme")

    params = _parse_auth_params(rest.lstrip())

    id_ = _require_param(params, "id")
    if not id_:
        raise ParseError("Empty 'id' parameter")

    realm = _require_param(params, "realm")
    method_raw = _require_param(params, "method")
    if not method_raw or not all(c.islower() and c.isascii() for c in method_raw):
        raise ParseError(f'Invalid method: "{method_raw}". Must be lowercase ASCII.')

    intent = _require_param(params, "intent")
    request_b64 = _require_param(params, "request")

    # Audit #9: cap the ``request`` parameter before base64url-decode + JSON
    # parse, matching ``parse_authorization`` / ``parse_receipt`` which cap the
    # token they decode. ``request`` is the only challenge field that drives
    # O(n) decode+parse cost; every other param is a short pass-through string.
    if len(request_b64) > MAX_TOKEN_LEN:
        raise ParseError(f"request parameter exceeds maximum length of {MAX_TOKEN_LEN} bytes")

    # Validate that request is valid base64url JSON
    try:
        request_bytes = decode(request_b64)
        json.loads(request_bytes)
    except Exception as exc:
        raise ParseError(f"Invalid JSON in request field: {exc}") from exc

    return PaymentChallenge(
        id=id_,
        realm=realm,
        method=method_raw,
        intent=intent,
        request=request_b64,
        expires=params.get("expires", ""),
        description=params.get("description", ""),
        digest=params.get("digest", ""),
        opaque=params.get("opaque"),
    )


def parse_www_authenticate_all(headers: Iterable[str]) -> list[PaymentChallenge]:
    """Parse all Payment challenges from one or more WWW-Authenticate values."""
    challenges: list[PaymentChallenge] = []
    for header in headers:
        for challenge in _extract_challenges(header, PAYMENT_SCHEME):
            try:
                challenges.append(parse_www_authenticate(challenge))
            except ParseError:
                continue
    return challenges


def format_www_authenticate(challenge: PaymentChallenge) -> str:
    """Format a PaymentChallenge as a WWW-Authenticate header value."""
    parts = [
        f'id="{_escape_quoted_value(challenge.id)}"',
        f'realm="{_escape_quoted_value(challenge.realm)}"',
        f'method="{_escape_quoted_value(challenge.method)}"',
        f'intent="{_escape_quoted_value(challenge.intent)}"',
        f'request="{_escape_quoted_value(challenge.request)}"',
    ]
    # Canonical mpp-tools order: description, digest, expires after request.
    # description round-trips as a top-level header param (parse keeps it,
    # format emits it) to match the canonical golden wire.
    if challenge.description:
        parts.append(f'description="{_escape_quoted_value(challenge.description)}"')
    if challenge.digest:
        parts.append(f'digest="{_escape_quoted_value(challenge.digest)}"')
    if challenge.expires:
        parts.append(f'expires="{_escape_quoted_value(challenge.expires)}"')
    if challenge.opaque is not None:
        parts.append(f'opaque="{_escape_quoted_value(challenge.opaque)}"')

    return f"Payment {', '.join(parts)}"


# ---------------------------------------------------------------------------
# Authorization
# ---------------------------------------------------------------------------


def parse_authorization(header: str) -> PaymentCredential:
    """Parse an Authorization header into a PaymentCredential."""
    payment_part = _extract_payment_scheme(header)
    if payment_part is None:
        raise ParseError("Expected 'Payment' scheme")

    token = payment_part[8:].strip() if len(payment_part) > 8 else ""

    if len(token) > MAX_TOKEN_LEN:
        raise ParseError(f"Token exceeds maximum length of {MAX_TOKEN_LEN} bytes")

    try:
        decoded = decode(token)
        data = json.loads(decoded)
    except Exception as exc:
        raise ParseError(f"Invalid credential: {exc}") from exc

    if not isinstance(data, dict) or "challenge" not in data:
        raise ParseError("Invalid credential JSON structure")

    ch = data["challenge"]
    if not isinstance(ch, dict):
        raise ParseError("Invalid credential challenge structure")
    # Canonical: the embedded challenge must carry a non-empty id.
    if not ch.get("id"):
        raise ParseError("Missing 'id' in credential challenge")
    echo = ChallengeEcho(
        id=str(ch.get("id", "")),
        realm=str(ch.get("realm", "")),
        method=str(ch.get("method", "")),
        intent=str(ch.get("intent", "")),
        request=str(ch.get("request", "")),
        expires=str(ch.get("expires", "")),
        digest=str(ch.get("digest", "")),
        opaque=ch.get("opaque"),
    )

    return PaymentCredential(
        challenge=echo,
        payload=data.get("payload", {}),
        source=data.get("source"),
    )


def format_authorization(credential: PaymentCredential) -> str:
    """Format a PaymentCredential as an Authorization header value."""
    challenge_dict: dict[str, Any] = {
        "challenge": {
            "id": credential.challenge.id,
            "intent": credential.challenge.intent,
            "method": credential.challenge.method,
            "realm": credential.challenge.realm,
            "request": credential.challenge.request,
        },
        "payload": credential.payload,
    }
    # Add optional challenge fields
    ch_dict = challenge_dict["challenge"]
    if credential.challenge.expires:
        ch_dict["expires"] = credential.challenge.expires
    if credential.challenge.digest:
        ch_dict["digest"] = credential.challenge.digest
    if credential.challenge.opaque is not None:
        ch_dict["opaque"] = credential.challenge.opaque

    if credential.source:
        challenge_dict["source"] = credential.source

    encoded = encode_json(challenge_dict)
    return f"Payment {encoded}"


# ---------------------------------------------------------------------------
# Payment-Receipt
# ---------------------------------------------------------------------------


def parse_receipt(header: str) -> Receipt:
    """Parse a Payment-Receipt header into a Receipt."""
    token = header.strip()
    if len(token) > MAX_TOKEN_LEN:
        raise ParseError(f"Receipt exceeds maximum length of {MAX_TOKEN_LEN} bytes")

    try:
        data = decode_json(token)
    except Exception as exc:
        raise ParseError(f"Invalid receipt: {exc}") from exc

    if not isinstance(data, dict):
        raise ParseError("Invalid receipt JSON structure")

    # Canonical: status / method / reference / timestamp are required.
    for field in ("status", "method", "reference", "timestamp"):
        if not data.get(field):
            raise ParseError(f"Missing '{field}' in receipt")

    timestamp = str(data["timestamp"])
    if not _ISO8601_RE.match(timestamp):
        raise ParseError(f"Invalid ISO-8601 timestamp in receipt: {timestamp!r}")

    return Receipt(
        status=str(data["status"]),
        method=str(data["method"]),
        timestamp=timestamp,
        reference=str(data["reference"]),
        challenge_id=str(data.get("challengeId", "")),
        external_id=str(data.get("externalId", "")),
    )


def format_receipt(receipt: Receipt) -> str:
    """Format a Receipt as a Payment-Receipt header value."""
    data: dict[str, Any] = {
        "challengeId": receipt.challenge_id,
        "method": receipt.method,
        "reference": receipt.reference,
        "status": receipt.status,
        "timestamp": receipt.timestamp,
    }
    if receipt.external_id:
        data["externalId"] = receipt.external_id
    return encode_json(data)


# ---------------------------------------------------------------------------
# Private helpers
# ---------------------------------------------------------------------------


def _strip_payment_scheme(header: str) -> str | None:
    """Strip the 'Payment' scheme prefix, returning the rest or None."""
    header = header.lstrip()
    if len(header) < len(PAYMENT_SCHEME):
        return None
    if header[: len(PAYMENT_SCHEME)].lower() != PAYMENT_SCHEME.lower():
        return None
    return header[len(PAYMENT_SCHEME) :]


def _extract_payment_scheme(header: str) -> str | None:
    """Extract the Payment scheme section from a comma-separated header."""
    for part in _extract_challenges(header, PAYMENT_SCHEME):
        return part
    return None


def _extract_challenges(header: str, scheme: str) -> list[str]:
    """Extract scheme challenges from a WWW-Authenticate header value."""
    all_starts = _find_challenge_starts(header)
    selected_starts = [start for start in all_starts if _scheme_starts_at(header, start, scheme)]
    challenges: list[str] = []
    for start in selected_starts:
        following_starts = [candidate for candidate in all_starts if candidate > start]
        end = following_starts[0] if following_starts else len(header)
        challenge = header[start:end].strip()
        if challenge.endswith(","):
            challenge = challenge[:-1].rstrip()
        if challenge:
            challenges.append(challenge)
    return challenges


def _find_challenge_starts(header: str) -> list[int]:
    starts: list[int] = []
    in_quote = False
    escaped = False
    i = 0
    while i < len(header):
        char = header[i]
        if escaped:
            escaped = False
            i += 1
            continue
        if char == "\\" and in_quote:
            escaped = True
            i += 1
            continue
        if char == '"':
            in_quote = not in_quote
            i += 1
            continue
        if not in_quote and _auth_scheme_starts_at(header, i):
            starts.append(i)
            while i < len(header) and header[i] not in (" ", "\t"):
                i += 1
            continue
        i += 1
    return starts


def _scheme_starts_at(header: str, index: int, scheme: str) -> bool:
    return header[index : index + len(scheme)].lower() == scheme.lower() and _auth_scheme_starts_at(header, index)


def _auth_scheme_starts_at(header: str, index: int) -> bool:
    if index >= len(header) or not header[index].isalpha():
        return False
    next_index = index
    while next_index < len(header) and header[next_index].isalpha():
        next_index += 1
    if next_index == index or next_index >= len(header) or header[next_index] not in (" ", "\t"):
        return False
    if index == 0:
        return True
    previous = index - 1
    while previous >= 0 and header[previous] in (" ", "\t"):
        previous -= 1
    return previous >= 0 and header[previous] == ","


def _escape_quoted_value(s: str) -> str:
    """Escape a string for use in a quoted-string header value."""
    if "\r" in s or "\n" in s:
        raise ParseError("Header value contains invalid CRLF characters")
    return s.replace("\\", "\\\\").replace('"', '\\"')


def _require_param(params: dict[str, str], key: str) -> str:
    """Require a parameter is present in the params dict."""
    if key not in params:
        raise ParseError(f"Missing '{key}' field")
    return params[key]


def _parse_auth_params(params_str: str) -> dict[str, str]:
    """Parse key=value or key="value" pairs from an auth-params string."""
    params: dict[str, str] = {}
    chars = list(params_str)
    i = 0

    while i < len(chars):
        # Skip whitespace and commas
        while i < len(chars) and (chars[i] in (" ", "\t", ",")):
            i += 1
        if i >= len(chars):
            break

        # Read key
        key_start = i
        while i < len(chars) and chars[i] != "=" and chars[i] not in (" ", "\t"):
            i += 1
        if i >= len(chars) or chars[i] != "=":
            # Skip to next comma or whitespace
            while i < len(chars) and chars[i] not in (" ", "\t", ","):
                i += 1
            continue

        key = "".join(chars[key_start:i])
        i += 1  # skip '='

        if i >= len(chars):
            break

        # Read value
        if chars[i] == '"':
            i += 1  # skip opening quote
            value_parts: list[str] = []
            while i < len(chars) and chars[i] != '"':
                if chars[i] == "\\" and i + 1 < len(chars):
                    i += 1
                    value_parts.append(chars[i])
                else:
                    value_parts.append(chars[i])
                i += 1
            if i < len(chars):
                i += 1  # skip closing quote
            value = "".join(value_parts)
        else:
            value_start = i
            while i < len(chars) and chars[i] not in (" ", "\t", ","):
                i += 1
            value = "".join(chars[value_start:i])

        if key in params:
            raise ParseError(f"Duplicate parameter: {key}")
        params[key] = value

    return params
